parlai/mturk/tasks/acute_eval/run.py

```diff
@@ -112,341 +102,379 @@ def add_args(from_argv=False):
+        tasks_per_hit = self.opt['subtasks_per_hit']
+        task_data = self.get_onboarding_tasks(worker_id)
+        if len(task_data) == tasks_per_hit:
+            return task_data
+        tries = 0
+
+        while (not self.task_queue.empty()) and tries < self.task_queue.qsize():
+            try:
+                next_task = self.task_queue.get()
+            except Queue.Empty:
+                break
+            tries += 1
+
+            pair_id = next_task['pair_id']
+            dialogue_ids = self._get_dialogue_ids(next_task)
+
+            if pair_id not in worker_data[  # make sure worker has not seen these conversations before
+                'tasks_completed'
+            ] and all(
+                d_id not in worker_data['conversations_seen'] for d_id in dialogue_ids
+            ):
+                # track tasks and conversations seen
+                worker_data['tasks_completed'].append(pair_id)
+                worker_data['conversations_seen'].extend(dialogue_ids)
+                task_data.append(next_task)
+                if len(task_data) == tasks_per_hit:
+                    return task_data
+            else:
+                self.task_queue.put(next_task)
+        # task queue containing num annotations requested of each pair is exhausted
+        # b/c we released enough hits to guarantee reaching the requested num on average
+        tasks_still_needed = tasks_per_hit - len(task_data)
+        tasks_remaining = [
+            t_id
+            for t_id in range(len(self.desired_tasks))
+            if t_id not in worker_data['tasks_completed']
+        ]
+        # get any pairings with conversations this worker has not seen to fill this hit
+        tasks_chosen = [
+            t
+            for t in tasks_remaining
+            if all(
+                d_id not in worker_data['conversations_seen']
+                for d_id in self._get_dialogue_ids(self.desired_tasks[t])
             )
-            while not world.episode_done():
-                world.parley()
-
-            world.shutdown()
-
-            save_data = world.prep_save_data(workers)
-
-            if not world.did_complete():
-                return_task_data(workers[0].worker_id, task_data)
-            elif opt['block_on_onboarding_fail']:
-                # check whether workers failed onboarding
-                check_and_update_worker_approval(
-                    mturk_manager,
-                    workers[0].worker_id,
-                    opt['onboarding_threshold'],
-                    save_data,
-                )
-            return save_data
+        ]
+        if tasks_still_needed < len(tasks_chosen):
+            tasks_chosen = random.sample(tasks_chosen, tasks_still_needed)
+        worker_data['tasks_completed'].extend(tasks_chosen)
+        for t in tasks_chosen:
+            worker_data['conversations_seen'].extend(
+                self._get_dialogue_ids(self.desired_tasks[t])
```

Comment:
do we need to add teh `pair_id` to `tasks_completed` here??

Reply:
yes, as it is possible that the worker hasn't actually seen this task (given that the `task_queue` is shared amongst all workers)